ltm/agent.py

```diff
@@ -1,102 +1,408 @@
+        response_text = ask_llm(context, model=self.model, max_overall_tokens=self.max_prompt_size, cost_callback=cost_callback, temperature=self.temperature)
+        debug_actions(context, self.temperature, response_text, self.llm_call_idx, self.debug_level, self.save_name,  name_template="reply-{idx}")
+        self.llm_call_idx += 1
+
+        # Save interaction to memory
+        um = Message(role="user", content=user_message, timestamp=self.now.timestamp())
+        am = Message(role="assistant", content=response_text, timestamp=self.now.timestamp())
+
+        self.save_interaction(user_message, response_text, keywords)
+        self.session.add_interaction((um, am))
+
+        return response_text
+
+    def keywords_for_message(self, user_message, cost_cb):
+
+        prompt = 'Create two keywords to describe the topic of this message:\n"{user_message}".\n\nFocus on the topic and tone of the message. Produce the keywords in JSON like: `["keyword_1", "keyword_2"]`\n\nChoose keywords that would aid in retriving this message from memory in the future.\n\nReuse these keywords if appropriate: {keywords}'
+
+        context = [make_system_message(prompt.format(user_message=user_message, keywords=self.defined_kws))]
+        while True:
+            try:
+                print("Keyword gen")
+                response = ask_llm(context, model=self.model, max_overall_tokens=self.max_prompt_size, cost_callback=cost_cb, temperature=self.temperature)
+
+                keywords = [k.lower() for k in sanitize_and_parse_json(response)]
+                break
+            except Exception as e:
+                print(repr(e) + response)
+                continue
+
+        # Update known list of keywords
+        for k in keywords:
+            if k not in self.defined_kws:
+                self.defined_kws.append(k)
+
+        print(f"Interaction keywords: {keywords}")
+        return keywords
+
+    def create_context(self, user_message, max_prompt_size, previous_interactions, cost_cb):
+
+        stamped_user_message = str(self.now) + ": " + user_message
+        context = [make_system_message(self.system_message), make_user_message(stamped_user_message)]
+        relevant_interactions = self.get_relevant_memories(user_message, cost_cb)
+
+        # Get interactions from the memories
+        for m in relevant_interactions:
+            if "trivia" in m[0].content:
+                colour_print("YELLOW", f"<*** trivia ***>")
+            else:
+                colour_print("YELLOW", f"{m[0].content}")
+
+        # Add the previous messages
+        final_idx = self.session.message_count - 1
+        while previous_interactions > 0 and final_idx > 0:
+
+            # Agent reply
+            context.insert(1, self.session.by_index(final_idx).as_llm_dict())
+            # User message
+            context.insert(1, self.session.by_index(final_idx-1).as_llm_dict())
```

Comment:
These messages are not timestamped. In the previous baseline they were timestamped. I guess you didn't notice it because you're not adding any to the context, but it seems right to continue the pattern.

Reply:
The actually are, just after they are added to the context see the lines below. We update the one at context[1], which is the user message.

Reply:
I see that all user messages before these interactions (added in the loop after this) are timestamped, but if `previous_interactions > 0`, the most recent interactions added in this loop don't have a timestamp. Am I missing something?

Reply:
Oh yeah, you are correct here, I will actually tidy it up and unify the two